### PR TITLE
cmake: emu: nsim: Allow all nsim platforms use mdb 

### PR DESCRIPTION
**Pull request by wangnuannuan**
Now the twister only supports `nsim_hs_smp` running application with nsim-mdb runner, this commit allows user run target with mdb-nsim runner. ~~What user need to do is to change the simulation type, for example https://github.com/wangnuannuan/zephyr/blob/mainfest-error/boards/arc/nsim/nsim_sem.yaml#L5~~

With this change cmake will choose runner according to `BOARD_DEBUG_RUNNER` or `BOARD_FLASH_RUNNER`. so if add `-DBOARD_FLASH_RUNNER=mdb-nsim` or `-DBOARD_DEBUG_RUNNER=mdb-nsim` to cmake command, the apllication will be run with mdb.

Signed-off-by: Jingru Wang <jingru@synopsys.com>

### CODE REVIEW

**Comment by abrodkin** [comment]
@wangnuannuan could you please add an example of how that might be used now with your change?

**Comment by wangnuannuan** [comment]
When the platform is nsim* in addition to nsim_hs_smp, the default runner is [arc-nsim](https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/arc/nsim/board.cmake#L4), in other words the platform run application with `nsimdrv`. [Here](https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/pylib/twister/twisterlib.py#L520) the twister doesn't specify the runner with `--runner`, so twister always runs application with `nsimdrv`. 

My change specific the runner as `mdb-nsim` when the platform's simulation is 'mdb-nsim'.

So if you want `twister` to run applcation with `mdb`, you only need to change the platform's simulation as `mdb-nsim`(in platform's yaml file), then the `twister` will choose `mdb-nsim` as runner

File: cmake/emu/nsim.cmake [hunk 1]
```diff
@@ -1,63 +1,95 @@
...
+    foreach(val RANGE ${CONFIG_MP_NUM_CPUS})
+      if(val LESS CONFIG_MP_NUM_CPUS)
+        MATH(EXPR PSET_NUM "${val}+1")
+        if (val GREATER 0)
+          list(APPEND MDB_OPTIONS &&)
+        endif()
+        list(APPEND MDB_OPTIONS ${MDB} -pset=${PSET_NUM} -psetname=core${val})
+        if (val GREATER 0)
+          list(APPEND MDB_OPTIONS -prop=download=2)
+          set(MULTIFILES ${MULTIFILES},core${val})
```
Comment (abrodkin) [inline cmake/emu/nsim.cmake:37]: See https://github.com/zephyrproject-rtos/zephyr/pull/36535 on how we order cores (`core0` to be added in the end - rationale is in @IRISZZW's PR).

File: cmake/emu/nsim.cmake [hunk 2]
```diff
@@ -1,63 +1,95 @@
 # SPDX-License-Identifier: Apache-2.0
-if(${CONFIG_SOC_NSIM_HS_SMP})
-# mdb is required to run nsim multicore targets
-find_program(
-  MDB
-  mdb
+if("${BOARD_DEBUG_RUNNER}" STREQUAL "mdb-nsim" OR "${BOARD_FLASH_RUNNER}" STREQUAL "mdb-nsim")
+  # mdb is required to run nsim multicore targets
```
Comment (abrodkin) [inline cmake/emu/nsim.cmake:3]: Move space changes to a separate commit, see more on that below.

**Review by abrodkin — CHANGES_REQUESTED** [review]
The idea itself is good, though see my comments on what might be improved.

File: cmake/emu/nsim.cmake [hunk 3]
```diff
@@ -1,26 +1,62 @@
...
+elseif(${CONFIG_SOC_NSIM_SEM})
+ set(MDB_ARGS mdb_sem.args)
+elseif(${CONFIG_SOC_NSIM_HS})
+ set(MDB_ARGS mdb_hs.args)
+elseif(${CONFIG_SOC_NSIM_HS6X})
+ set(MDB_ARGS mdb_hs6x.args)
+endif()
+
+# remove previous .sc.project folder which has temporary settings for MDB.
+set(MDB_OPTIONS rm -rf ${APPLICATION_BINARY_DIR}/.sc.project)
```
Comment (evgeniy-paltsev) [inline cmake/emu/nsim.cmake:25]: I'm curious if it is really working on Windows platform?

File: cmake/emu/nsim.cmake [hunk 4]
```diff
@@ -1,26 +1,62 @@
...
+ set(MDB_ARGS mdb_em.args)
+elseif(${CONFIG_SOC_NSIM_EM7D_V22})
+ set(MDB_ARGS mdb_em7d_v22.args)
+elseif(${CONFIG_SOC_NSIM_SEM})
+ set(MDB_ARGS mdb_sem.args)
+elseif(${CONFIG_SOC_NSIM_HS})
+ set(MDB_ARGS mdb_hs.args)
+elseif(${CONFIG_SOC_NSIM_HS6X})
+ set(MDB_ARGS mdb_hs6x.args)
+endif()
```
Comment (evgeniy-paltsev) [inline cmake/emu/nsim.cmake:22]: Should we move it to `boards/arc/nsim/board.cmake`?

We do similar thing for runners arguments in it:
```
if(${CONFIG_SOC_NSIM_EM})
board_runner_args(arc-nsim "--props=nsim_em.props")
board_runner_args(mdb-nsim "--nsim_args=mdb_em.args")
elseif(${CONFIG_SOC_NSIM_EM7D_V22})
board_runner_args(arc-nsim "--props=nsim_em7d_v22.props")
board_runner_args(mdb-nsim "--nsim_args=mdb_em7d_v22.args")
elseif(${CONFIG_SOC_NSIM_SEM})
board_runner_args(arc-nsim "--props=nsim_sem.props")
board_runner_args(mdb-nsim "--nsim_args=mdb_sem.args")
elseif(${CONFIG_SOC_NSIM_HS})
board_runner_args(arc-nsim "--props=nsim_hs.props")
 [**snip**]
```

File: boards/arc/nsim/board.cmake [hunk 5]
```diff
@@ -8,21 +8,32 @@ endif()
 
 if(${CONFIG_SOC_NSIM_EM})
 board_runner_args(arc-nsim "--props=nsim_em.props")
+set(NSIM_PROPS nsim_em.props)
 board_runner_args(mdb-nsim "--nsim_args=mdb_em.args")
+set(MDB_ARGS mdb_hs_smp.args)
```
Comment (ruuddw) [inline boards/arc/nsim/board.cmake:13]: With the NSIM_PROPS and MDB_ARGS variables, I suggest to factor out the 
    board_runner_args(arc-nsim "--props=nsim_em.props")
    board_runner_args(mdb-nsim "--nsim_args=mdb_em.args")

and replace them with a single copy using the variables. Something like
    board_runner_args(arc-nsim "--props=${NSIM_PROPS }")
    board_runner_args(mdb-nsim "--nsim_args="${MBD_ARGS}")

Comment (ruuddw) [reply boards/arc/nsim/board.cmake]: I like the change for this, makes it even more compact and scalable, instead of all the hardcoded board names.

**Review by ruuddw — APPROVED** [review]

File: cmake/emu/nsim.cmake [hunk 6]
```diff
@@ -1,44 +1,54 @@
...
-  ${MDB} -pset=2 -psetname=core1 -prop=download=2 -prop=ident=0x00000150 -cmpd=soc
-  @${BOARD_DIR}/support/${MDB_ARGS} ${APPLICATION_BINARY_DIR}/zephyr/${KERNEL_ELF_NAME} &&
-  NSIM_MULTICORE=1 ${MDB} -multifiles=core0,core1 -cmpd=soc -run -cl
+  ${MDB_OPTIONS}
+  ${APPLICATION_BINARY_DIR}/zephyr/${KERNEL_ELF_NAME}
   DEPENDS ${logical_target_for_zephyr_elf}
   WORKING_DIRECTORY ${APPLICATION_BINARY_DIR}
   USES_TERMINAL
-  )
+)
```
Comment (evgeniy-paltsev) [inline cmake/emu/nsim.cmake:45]: It would be nice to add printing of MDB command here (i.e. if verbose mode is enabled).
As we construct the command in a quite complex way such logging will allow user to easily get this command to inspect it or run it manually.

Comment (evgeniy-paltsev) [inline cmake/emu/nsim.cmake:45]: It would be nice to add printing of MDB command here (i.e. if verbose mode is enabled).
As we construct the command in a quite complex way such logging will allow user to easily get this command to inspect it or run it manually.

**Review by evgeniy-paltsev — CHANGES_REQUESTED** [review]

File: cmake/emu/nsim.cmake (hunk 6 above)
Comment (wangnuannuan) [reply cmake/emu/nsim.cmake]: I add command to `COMMENT` which is the option of `add_custom_target`. When you run the command `make run` or `ninja run`, it will display the command.

**Review by abrodkin — APPROVED** [review]

**Comment by evgeniy-paltsev** [comment]
Hi Jingru,

I've tried to test your changes, however it looks like we have some issues with them.
When I'm building for nsim_hs_smp the `west flash` command is working, but ninja run isn't.

How to reproduce:
```
west build -b nsim_hs_smp samples/hello_world
west -v flash
 [**snip**]
 [zephyr print] *** Booting Zephyr OS build v2.6.0-rc3-1967-g70458aeb5bf8  ***
 [zephyr print] Hello World! nsim

cd build/
ninja -v run
 [**snip**]
 There is no printing from zephyr!
```

Note that ninja is used by default (instead of make).
Zephyr SDK is 0.13.0-alfa

File: boards/arc/nsim/board.cmake [hunk 7]
```diff
@@ -4,27 +4,16 @@ set(EMU_PLATFORM nsim)
...
-elseif(${CONFIG_SOC_NSIM_HS_SMP})
-board_runner_args(mdb-nsim "--cores=${CONFIG_MP_NUM_CPUS}" "--nsim_args=mdb_hs_smp.args")
-elseif(${CONFIG_SOC_NSIM_HS6X})
-board_runner_args(arc-nsim "--props=nsim_hs6x.props")
-board_runner_args(mdb-nsim "--nsim_args=mdb_hs6x.args")
-elseif(${CONFIG_SOC_NSIM_HS6X_SMP})
-board_runner_args(mdb-nsim "--cores=${CONFIG_MP_NUM_CPUS}" "--nsim_args=mdb_hs6x_smp.args")
+string(REPLACE "nsim" "mdb" MDB_ARGS "${BOARD}.args")
+if((CONFIG_SOC_NSIM_HS_SMP OR CONFIG_SOC_NSIM_HS6X_SMP))
+board_runner_args(mdb-nsim "--cores=${CONFIG_MP_NUM_CPUS}" "--nsim_args=${MDB_ARGS}")
```
Comment (evgeniy-paltsev) [inline boards/arc/nsim/board.cmake:14]: Just a nitpick: I guess we should add proper formatting here (indentation)

File: boards/arc/nsim/board.cmake [hunk 8]
```diff
@@ -4,27 +4,16 @@ set(EMU_PLATFORM nsim)
...
+else()
+board_runner_args(mdb-nsim "--nsim_args=${MDB_ARGS}")
```
Comment (evgeniy-paltsev) [inline boards/arc/nsim/board.cmake:16]: Same here.

File: cmake/emu/nsim.cmake [hunk 9]
```diff
@@ -1,46 +1,54 @@
...
- set(MDB_ARGS mdb_hs6x_smp.args)
+# remove previous .sc.project folder which has temporary settings for MDB.
+set(MDB_OPTIONS ${CMAKE_COMMAND} -E rm -rf ${APPLICATION_BINARY_DIR}/.sc.project)
+if(CONFIG_MP_NUM_CPUS GREATER 1)
+  set(MULTIFILES ${MDB} -multifiles=)
+  foreach(val RANGE ${CONFIG_MP_NUM_CPUS})
+    if(val LESS CONFIG_MP_NUM_CPUS)
+      MATH(EXPR PSET_NUM "${CONFIG_MP_NUM_CPUS}-${val}")
+      MATH(EXPR CORE_NUM "${CONFIG_MP_NUM_CPUS}-${val}-1")
+      if (PSET_NUM GREATER 0)
```
Comment (evgeniy-paltsev) [inline cmake/emu/nsim.cmake:18]: As I remember we should shouldn't add extra space between `if` and `(` in cmake.

File: cmake/emu/nsim.cmake [hunk 10]
```diff
@@ -1,46 +1,54 @@
...
+        list(APPEND MDB_OPTIONS &&)
+      endif()
+      list(APPEND MDB_OPTIONS ${MDB} -pset=${PSET_NUM} -psetname=core${CORE_NUM})
+      if (PSET_NUM GREATER 1)
```
Comment (evgeniy-paltsev) [inline cmake/emu/nsim.cmake:22]: same here.

**Review by evgeniy-paltsev — COMMENTED** [review]
Hi Jingru,
several nitpicks from my side :)

File: boards/arc/nsim/board.cmake [hunk 11]
```diff
@@ -2,29 +2,18 @@
 set(EMU_PLATFORM nsim)
 
 if(NOT (CONFIG_SOC_NSIM_HS_SMP OR CONFIG_SOC_NSIM_HS6X_SMP))
-board_set_flasher_ifnset(arc-nsim)
+  board_set_flasher_ifnset(arc-nsim)
 board_set_debugger_ifnset(arc-nsim)
+
+board_runner_args(arc-nsim "--props=${BOARD}.props")
+set(NSIM_PROPS "${BOARD}.props")
```
Comment (evgeniy-paltsev) [inline boards/arc/nsim/board.cmake:9]: Just a nitpick: please fix formatting here (indentation) for the whole if block.

**Review by evgeniy-paltsev — COMMENTED** [review]
Hi Jingru,

I've re-tested it and now it works fine, thanks. Please find one minor comment, rest of the fings looks good.

**Review by evgeniy-paltsev — APPROVED** [review]
